### PR TITLE
JAMES-3953 Provide a file based blob store

### PR DESCRIPTION
**Pull request by chibenwa**
See the ticket for details but in short having such a card in the James game would ease integrating James into legacy storage systems, not cloud / s3 ready. Think ISCI, SAN, NAS and distributed systems like CEPH...

### CODE REVIEW

File: server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        Preconditions.checkNotNull(inputStream);
+        File bucketRoot = getBucketRoot(bucketName);
+        File blob = new File(bucketRoot, blobId.asString());
+        return Mono.fromRunnable(() -> {
+                save(inputStream, blob);
+            })
```
Comment (Arsnael) [inline server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java:116]:
```
return Mono.fromRunnable(() -> save(inputStream, blob))
```

File: server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            .subscribeOn(Schedulers.boundedElastic())
+            .then()
+            .retryWhen(Retry.backoff(10, Duration.ofMillis(100))
+                .filter(e -> e instanceof OverlappingFileLockException));
+    }
+
+    private void save(InputStream inputStream, File blob) {
+        if (blob.exists()) {
+            return;
+        }
```
Comment (Arsnael) [inline server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java:126]: line space for readability

File: server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        try (FileOutputStream out = new FileOutputStream(blob);
+             FileChannel channel = out.getChannel();
+             FileLock fileLock = channel.lock()) {
+            inputStream.transferTo(out);
+
```
Comment (Arsnael) [inline server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreDAO.java:131]: unnecessary empty line

File: server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreFactory.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    private final FileSystem fileSystem;
+
+    public FileBlobStoreFactory(FileSystem fileSystem) {
+        this.fileSystem = fileSystem;
+    }
+
+    public BlobStoreFactory.RequireBlobIdFactory builder() {
+        return blobIfFactory -> {
+            try {
+                return BlobStoreFactory.builder().blobStoreDAO(new FileBlobStoreDAO(fileSystem, blobIfFactory))
```
Comment (Arsnael) [inline server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreFactory.java:37]:
```
return BlobStoreFactory.builder()
    .blobStoreDAO(new FileBlobStoreDAO(fileSystem, blobIfFactory))
```

File: server/container/guice/distributed/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 5]
```diff
@@ -75,6 +76,15 @@ protected void configure() {
         }
     }
 
+    static class FileBlobStoreDAODeclarationModule extends AbstractModule {
+        @Override
+        protected void configure() {
+            install(new CassandraBucketModule());
```
Comment (Arsnael) [inline server/container/guice/distributed/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:82]: Why a Cassandra module?

File: server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreFactory.java (hunk 4 above)
Comment (quantranhong1999) [reply server/blob/blob-file/src/main/java/org/apache/james/blob/file/FileBlobStoreFactory.java:37]: `blobIfFactory` -> `blobIdFactory`

**Review by quantranhong1999 — APPROVED** [review]

**Comment by Arsnael** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1766/2/

Failed tests related

**Comment by Arsnael** [comment]
```
21:15:52,002 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.0:check (check-style) on project scaling-pulsar-smtp: You have 1 Checkstyle violation. -> [Help 1]
```
And next time fixups would be appreciated so we don't need to review the all PR again :)

**Review by Arsnael — APPROVED** [review]

**Pull request merged**